File: src/main/java/svnserver/repository/RepositoryMapping.java
/*
 * This file is part of git-as-svn. It is subject to the license terms
 * in the LICENSE file found in the top-level directory of this distribution
 * and at http://www.gnu.org/licenses/gpl-2.0.html. No part of git-as-svn,
 * including this file, may be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file.
 */
package svnserver.repository;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.tmatesoft.svn.core.SVNErrorCode;
import org.tmatesoft.svn.core.SVNErrorMessage;
import org.tmatesoft.svn.core.SVNException;
import org.tmatesoft.svn.core.SVNURL;
import svnserver.StringHelper;
import svnserver.context.Shared;
import svnserver.parser.SvnServerWriter;
import svnserver.repository.git.BranchProvider;
import svnserver.repository.git.GitBranch;
import svnserver.server.command.BaseCmd;

import java.io.IOException;
import java.util.Map;
import java.util.NavigableMap;

/**
 * Resolving repository by URL.
 *
 * @author Artem V. Navrotskiy <bozaro@users.noreply.github.com>
 */
public interface RepositoryMapping<T extends BranchProvider> extends Shared {
  @Nullable
  static <T extends BranchProvider> RepositoryInfo findRepositoryInfo(@NotNull RepositoryMapping<T> mapping, @NotNull SVNURL url, @NotNull SvnServerWriter writer) throws SVNException, IOException {
    final String path = StringHelper.normalizeDir(url.getPath());
    final Map.Entry<String, T> repo = getMapped(mapping.getMapping(), path);
    if (repo == null) {
      BaseCmd.sendError(writer, SVNErrorMessage.create(SVNErrorCode.RA_SVN_REPOS_NOT_FOUND, "Repository not found: " + url));
      return null;
    }

    final String branchPath = repo.getKey().isEmpty() ? path : path.substring(repo.getKey().length() - 1);
    final NavigableMap<String, GitBranch> branches = repo.getValue().getBranches();

    if (branchPath.length() <= 1) {
      final String branchName = repo.getValue().getBranches().size() == 1
          ? repo.getValue().getBranches().values().iterator().next().getShortBranchName()
          : "<branchname>";
      final String msg = String.format("Repository branch not found. Use `svn relocate %s/%s` to fix your working copy", url, branchName);
      BaseCmd.sendError(writer, SVNErrorMessage.create(SVNErrorCode.RA_SVN_REPOS_NOT_FOUND, msg));
      return null;
    }

    final Map.Entry<String, GitBranch> branch = getMapped(branches, branchPath);

    if (branch == null) {
      BaseCmd.sendError(writer, SVNErrorMessage.create(SVNErrorCode.RA_SVN_REPOS_NOT_FOUND, "Repository branch not found: " + url));
      return null;
    }

    return new RepositoryInfo(
        SVNURL.create(
            url.getProtocol(),
            url.getUserInfo(),
            url.getHost(),
            url.getPort() == SVNURL.getDefaultPortNumber(url.getProtocol()) ? -1 : url.getPort(),
            repo.getKey() + branch.getKey().substring(1),
            true
        ),
        branch.getValue()
    );
  }

  @Nullable
  static <T> Map.Entry<String, T> getMapped(@NotNull NavigableMap<String, T> mapping, @NotNull String prefix) {
    final String path = StringHelper.normalizeDir(prefix);

    // TODO: this could be must faster if we find an appropriate trie implementation.
    Map.Entry<String, T> result = null;
    for (Map.Entry<String, T> entry : mapping.headMap(path, true).entrySet()) {
      if (!StringHelper.isParentPath(entry.getKey(), path))
        continue;

      if (result == null || entry.getKey().length() > result.getKey().length())
        result = entry;
    }

    return result;
  }

  @NotNull
  NavigableMap<String, T> getMapping();
}

File: src/main/kotlin/svnserver/repository/RepositoryMapping.kt
/*
 * This file is part of git-as-svn. It is subject to the license terms
 * in the LICENSE file found in the top-level directory of this distribution
 * and at http://www.gnu.org/licenses/gpl-2.0.html. No part of git-as-svn,
 * including this file, may be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file.
 */
package svnserver.repository

import org.tmatesoft.svn.core.SVNErrorCode
import org.tmatesoft.svn.core.SVNErrorMessage
import org.tmatesoft.svn.core.SVNException
import org.tmatesoft.svn.core.SVNURL
import svnserver.StringHelper
import svnserver.context.Shared
import svnserver.parser.SvnServerWriter
import svnserver.repository.git.BranchProvider
import svnserver.repository.git.GitBranch
import svnserver.server.command.BaseCmd
import java.io.IOException
import java.util.*

/**
 * Resolving repository by URL.
 *
 * @author Artem V. Navrotskiy <bozaro@users.noreply.github.com>
 */
interface RepositoryMapping<T : BranchProvider> : Shared {
    val mapping: NavigableMap<String, T>

    companion object {
        @Throws(SVNException::class, IOException::class)
        fun <T : BranchProvider> findRepositoryInfo(mapping: RepositoryMapping<T>, url: SVNURL, writer: SvnServerWriter): RepositoryInfo? {
            val path: String = StringHelper.normalizeDir(url.path)
            val repo: Map.Entry<String, T>? = getMapped(mapping.mapping, path)
            if (repo == null) {
                BaseCmd.sendError(writer, SVNErrorMessage.create(SVNErrorCode.RA_SVN_REPOS_NOT_FOUND, "Repository not found: $url"))
                return null
            }
            val branchPath: String = if (repo.key.isEmpty()) path else path.substring(repo.key.length - 1)
            val branches: NavigableMap<String, GitBranch> = repo.value.branches
            if (branchPath.length <= 1) {
                val branchName: String = if (repo.value.branches.size == 1) repo.value.branches.values.iterator().next().shortBranchName else "<branchname>"
                val msg: String = String.format("Repository branch not found. Use `svn relocate %s/%s` to fix your working copy", url, branchName)
                BaseCmd.sendError(writer, SVNErrorMessage.create(SVNErrorCode.RA_SVN_REPOS_NOT_FOUND, msg))
                return null
            }
            val branch: Map.Entry<String, GitBranch>? = (getMapped(branches, branchPath))!!
            if (branch == null) {
                BaseCmd.sendError(writer, SVNErrorMessage.create(SVNErrorCode.RA_SVN_REPOS_NOT_FOUND, "Repository branch not found: $url"))
                return null
            }
            return RepositoryInfo(
                SVNURL.create(
                    url.protocol,
                    url.userInfo,
                    url.host,
                    if (url.port == SVNURL.getDefaultPortNumber(url.protocol)) -1 else url.port, repo.key + branch.key.substring(1),
                    true
                ),
                branch.value
            )
        }

        fun <T> getMapped(mapping: NavigableMap<String, T>, prefix: String): Map.Entry<String, T>? {
            val path: String = StringHelper.normalizeDir(prefix)

            // TODO: this could be must faster if we find an appropriate trie implementation.
            var result: Map.Entry<String, T>? = null
            for (entry: Map.Entry<String, T> in mapping.headMap(path, true).entries) {
                if (!StringHelper.isParentPath(entry.key, path)) continue
                if (result == null || entry.key.length > result.key.length) result = entry
            }
            return result
        }
    }
}
